domain/include/cstone/traversal/groups_gpu.cu

```diff
@@ -0,0 +1,159 @@
+    DeviceVector<LocalIndex>& numSplitsPerGroup,
+    DeviceVector<LocalIndex>& groups)
+{
+    LocalIndex numParticles   = last - first;
+    LocalIndex numFixedGroups = iceil(numParticles, groupSize);
+    unsigned numThreads       = 256;
+    unsigned gridSize         = numFixedGroups * GpuConfig::warpSize;
+
+    splitMasks.resize(numFixedGroups);
+
+    numSplitsPerGroup.reserve(numFixedGroups * 1.1);
+    numSplitsPerGroup.resize(numFixedGroups);
+
+    groupSplitsKernel<groupSize><<<iceil(gridSize, numThreads), numThreads>>>(
+        first, last, x, y, z, h, leaves, numLeaves, layout, box, tolFactor, rawPtr(splitMasks),
+        rawPtr(numSplitsPerGroup), numFixedGroups);
+
+    groups.reserve(numFixedGroups * 1.1);
+    groups.resize(numFixedGroups + 1);
+    exclusiveScanGpu(rawPtr(numSplitsPerGroup), rawPtr(numSplitsPerGroup) + numFixedGroups + 1, rawPtr(groups));
+    LocalIndex newNumGroups;
+    memcpyD2H(rawPtr(groups) + groups.size() - 1, 1, &newNumGroups);
+
+    auto& newGroupSizes = numSplitsPerGroup;
+    newGroupSizes.resize(newNumGroups);
+
+    makeSplitsKernel<<<numFixedGroups, numThreads>>>(rawPtr(splitMasks), rawPtr(groups), numFixedGroups,
+                                                     rawPtr(newGroupSizes));
+
+    groups.resize(newNumGroups + 1);
+    exclusiveScanGpu(rawPtr(newGroupSizes), rawPtr(newGroupSizes) + newNumGroups + 1, rawPtr(groups), first);
+    memcpyH2D(&last, 1, rawPtr(groups) + groups.size() - 1);
+}
+
+template<class Tc, class T, class KeyType>
+void computeGroupSplits(LocalIndex first,
+                        LocalIndex last,
+                        const Tc* x,
+                        const Tc* y,
+                        const Tc* z,
+                        const T* h,
+                        const KeyType* leaves,
+                        TreeNodeIndex numLeaves,
+                        const LocalIndex* layout,
+                        const Box<Tc> box,
+                        unsigned groupSize,
+                        float tolFactor,
+                        DeviceVector<LocalIndex>& numSplitsPerGroup,
+                        DeviceVector<LocalIndex>& groups)
+{
+    if (groupSize == GpuConfig::warpSize)
+    {
+        DeviceVector<util::array<GpuConfig::ThreadMask, 1>> splitMasks;
+        computeGroupSplitsImpl<GpuConfig::warpSize>(first, last, x, y, z, h, leaves, numLeaves, layout, box, tolFactor,
+                                                    splitMasks, numSplitsPerGroup, groups);
+    }
+    else if (groupSize == 2 * GpuConfig::warpSize)
+    {
+        DeviceVector<util::array<GpuConfig::ThreadMask, 2>> splitMasks;
+        computeGroupSplitsImpl<2 * GpuConfig::warpSize>(first, last, x, y, z, h, leaves, numLeaves, layout, box,
+                                                        tolFactor, splitMasks, numSplitsPerGroup, groups);
+    }
+    else { throw std::runtime_error("Unsupported spatial group size\n"); }
```

Comment:
on x-daint, I get:
```
groups_gpu.cu(146): error: namespace "std" has no member "runtime_error"
```
with gcc/11.2, cuda/11.6 (stdexcept)

Reply:
should be fixed now